wallet-new/src/Cardano/Wallet/Kernel/DB/Spec/Update.hs

```diff
@@ -53,15 +61,70 @@ deriveSafeCopy 1 'base ''NewPendingFailed
 --   a specialized hardware device).
 -- * We do not actually have access to the key storage inside the DB layer
 --   (and do not store private keys) so we cannot actually sign transactions.
-newPending :: InDb (Core.TxAux)
+newPending :: InDb Core.TxAux
            -> Update' Checkpoints NewPendingFailed ()
-newPending _tx = error "newPending"
+newPending tx = do
+    checkpoints <- get
+    let available' = available (checkpoints ^. currentUtxo) (checkpoints ^. currentPendingTxs)
+    if isValidPendingTx tx' available'
+        then
+            put (insertPending checkpoints)
+        else
+            inputUnavailableErr available'
 
--- | Apply a block to /all/ wallets' UTxO.
---
--- Ideally the block is prefiltered (see 'prefilter').
-applyBlock :: (ResolvedBlock, BlockMeta) -> Checkpoints -> Checkpoints
-applyBlock (_resolvedBlock, _blockMeta) = error "applyBlock"
+    return ()
+    where
+        tx' = tx ^. fromDb
+
+        insertPending :: Checkpoints -> Checkpoints
+        insertPending cs = cs & currentPendingTxs %~ Map.insert txId tx'
+            where txId = hash $ Core.taTx tx'
+
+        inputUnavailableErr available_ = do
+            let unavailableInputs = txAuxInputSet tx' `Set.difference` utxoInputs available_
+            throwError $ NewPendingInputsUnavailable unavailableInputs
+
+-- | Apply the prefiltered block to the specified wallet
+applyBlock :: (PrefilteredBlock, BlockMeta)
+           -> Checkpoints
+           -> Checkpoints
+applyBlock (prefBlock, _bMeta) checkpoints
+    = Checkpoint {
+          _checkpointUtxo           = InDb utxo''
+        , _checkpointUtxoBalance    = InDb balance''
+        , _checkpointPending        = Pending . InDb $ pending''
+        , _checkpointExpected       = InDb expected''
+        , _checkpointBlockMeta      = blockMeta''
+        } NE.<| checkpoints
+    where
+        utxo'        = checkpoints ^. currentUtxo
+        utxoBalance' = checkpoints ^. currentUtxoBalance
+        pending'     = checkpoints ^. currentPendingTxs
+
+        (utxo'', balance'') = updateUtxo prefBlock (utxo', utxoBalance')
+        pending''           = updatePending prefBlock pending'
+        -- TODO applyBlock.updateExpected
+        expected''          = checkpoints ^. currentExpected
+        -- TODO applyBlock.updateBlockMeta
+        blockMeta''         = checkpoints ^. currentBlockMeta
```

Comment:
Yeah, so here also (block meta). Let's think about this before we think about rollback.